ENH: Support k!=0 for sparse.diagonal()

jaimefrio opened the pull request.

jaimefrio commented:
My editor also removed a lot of trailing white space, so the relevant diff is much smaller than the stats for csr.h and csc.h tell.

It still needs tests for the new functionality, but I need to figure out the sparse testing framework beforing tackling that.

I'd appreciate a review of the code in its current state, and any pointers on what to test and where to put those tests.

rgommers commented on scipy/sparse/sparsetools/csr.h:
```diff
@@ -28,30 +28,37 @@
+                  const I n_row,
+                  const I n_col,
+	              const I Ap[],
+	              const I Aj[],
```
if these lines are changed anyway, then maybe fix the inconsistent indentation as well?

rgommers commented on scipy/sparse/sparsetools/csr.h:
```diff
@@ -93,9 +100,9 @@ void expandptr(const I n_row,
-	                const I Aj[], 
+                    const I n_col,
+	                const I Ap[],
+	                const I Aj[],
```
more inconsistent indentation

rgommers commented on scipy/sparse/sparsetools/csr.h:
```diff
@@ -115,9 +122,9 @@ void csr_scale_rows(const I n_row,
-	                   const I Aj[], 
+                       const I n_col,
+	                   const I Ap[],
+	                   const I Aj[],
```
more inconsistent indentation

rgommers commented:
TravisCI is not yet happy. One style issue:

    $ pep8 scipy benchmarks/benchmarks
    scipy/sparse/compressed.py:524:40: E225 missing whitespace around operator
    1       E225 missing whitespace around operator

and just restarted the job that timed out

rgommers commented:
> any pointers on what to test and where to put those tests.

I think adding to ``_TestCommon.test_diagonal`` in ``test_base.py``. 

Testing valid and invalid values of ints (int32 and int64 maybe), for various shapes matrices, plus invalid inputs (floats, sequences, etc.) would seem reasonable to me.

perimosocordiae commented on scipy/sparse/base.py:
```diff
@@ -1019,11 +1019,13 @@ def _is_integral(dtype):
-        return self.tocsr().diagonal()
+        M, N = self.shape
+        if (k > 0 and k >= N) or (k < 0 and -k >= M):
+            raise ValueError("k exceeds matrix dimensions")
```
Right now `k` is validated twice for non-compressed sparse formats. I'd remove this check and let the implementing class handle it.

jaimefrio replied:
Done!

perimosocordiae commented on scipy/sparse/compressed.py:
```diff
@@ -513,13 +513,17 @@ def _mul_sparse_matrix(self, other):
-    def diagonal(self):
-        """Returns the main diagonal of the matrix
+    def diagonal(self, k=0):
+        """Returns the k-th diagonal of the matrix
```
Instead of duplicating the docstring here, we should use the docstring from `spmatrix`. ([example](https://github.com/scipy/scipy/blob/9597f6a0b5d415d65f8ffa5ab30f32d7a3b74ca0/scipy/sparse/compressed.py#L115))

jaimefrio replied:
Done!

perimosocordiae commented on scipy/sparse/compressed.py:
```diff
@@ -513,13 +513,17 @@ def _mul_sparse_matrix(self, other):
-        y = np.empty(min(self.shape), dtype=upcast(self.dtype))
-        fn(self.shape[0], self.shape[1], self.indptr, self.indices, self.data, y)
+        y = np.empty(min(M + (k if k < 0 else 0),
+                         N - (k if k >= 0 else 0)),
```
`M + min(k, 0)` and `N - max(k, 0)` seem a little easier to understand.

jaimefrio replied:
With the added benefit that they now fit in a single line. Done!

jaimefrio commented:
This turned into much more work than I was expecting, but arbitrary diagonals now work for all formats, including COO, DIA and BSR. Especially BSR was much more thinking than I was hoping to do on a Sunday afternoon...

Not sure how self explanatory the BSR code is, am willing to comment it in detail if needed. But not only does it handle arbitrary `k`, it is also more efficient for the general case. So much so that I have removed the special path that used to be there for the case `R == C`.

There are also a lot of formatting changes that happened when my editor converted all tabs to spaces in `bsr.h`, have gone ahead and fixed the indentation throughout as well.

Anyway, I think this is mostly ready now.

perimosocordiae commented on scipy/sparse/bsr.py:
```diff
@@ -293,17 +293,22 @@ def __repr__(self):
-    def diagonal(self):
+    def diagonal(self, k=0):
         """Returns the main diagonal of the matrix
         """
```
We're overwriting the docstring later, so we should remove it here.

jaimefrio commented:
Have added some comments to the BSR code and removed the extra docstrings.

perimosocordiae commented on scipy/sparse/base.py:
```diff
@@ -1019,11 +1019,10 @@ def _is_integral(dtype):
-    def diagonal(self):
+    def diagonal(self, k=0):
         """Returns the main diagonal of the matrix
         """
```
We're no longer returning just the main diagonal. This docstring should also define k, and maybe give a usage example? A see-also reference to numpy's diagonal would be good, too.

jaimefrio replied:
Oh yes, the docstring... Working on it, what's the next planned version? 0.20?

rgommers replied:
1.0 !

jaimefrio replied:
Yay! :-)

jaimefrio commented:
Docstring added before heading to bed, more tomorrow.

Not sure about how to specify a 'See also' entry referring to numpy, comments welcome.

Also, is this worthy of including in the release notes?

rgommers commented:
> Also, is this worthy of including in the release notes?

Borderline, but I'd say yes. And easier to put all enhancements into the release notes with the PRs and later edit down than having to hunt for things to put in the release notes at the last moment.

rgommers commented:
> Not sure about how to specify a 'See also' entry referring to numpy, comments welcome.

This will do it:

    See Also
    --------
    numpy.funcname

The magic of intersphinx:)

jaimefrio commented:
Have added a mention in in the release notes.

I think this is ready for a final review.

perimosocordiae commented:
Merged. The TravisCI failure was a timeout during coverage analysis, after having passed the test suite.

Thanks @jaimefrio !
